org.osgi.test.common/src/main/java/org/osgi/test/common/event/EventRecording.java

```diff
@@ -0,0 +1,193 @@
+/*******************************************************************************
+ * Copyright (c) Contributors to the Eclipse Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *******************************************************************************/
+
+package org.osgi.test.common.event;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.BundleEvent;
+import org.osgi.framework.BundleListener;
+import org.osgi.framework.FrameworkEvent;
+import org.osgi.framework.FrameworkListener;
+import org.osgi.framework.ServiceEvent;
+import org.osgi.framework.ServiceListener;
+import org.osgi.framework.SynchronousBundleListener;
+
+public interface EventRecording {
+	boolean isTimedOut();
+
+	List<TimedEvent<?>> timedEvents();
+
+	@SuppressWarnings("unchecked")
+	default <T> List<TimedEvent<T>> timedEvents(Class<T> clazz) {
+		return timedEvents().stream()
+			.filter((te) -> clazz.isInstance(te.getEvent()))
+			.map((te) -> (TimedEvent<T>) te)
```

Comment:
This method is a no-op. You want to assign the filtered collection to a variable of `List<TimedEvent<T>>` and make the case on that assignment.

```java
List<TimedEvent<T>> result = (List<TimedEvent<T>>) timedEvents().stream()
	.filter((te) -> clazz.isInstance(te.getEvent()))
	.collect(Collectors.toList());
return result;
```

Reply:
this would result in a compile error
```
Cannot cast from List<EventRecording.TimedEvent<?>> to List<EventRecording.TimedEvent<T>>
```

Reply:
So you may need to cast to a raw List first. And also suppress the raw warning.

```java
List<TimedEvent<T>> result = (List<TimedEvent<T>>)(List) timedEvents().stream()
	.filter((te) -> clazz.isInstance(te.getEvent()))
	.collect(Collectors.toList());
return result;
```

Reply:
double cast.. first time in my life!